src/System.Private.Interpreter/src/Internal/Runtime/Interpreter/ILInterpreter.cs

```diff
@@ -2126,5 +2123,321 @@ private void InterpretSwitch(ref ILReader reader, int jmpBase, int[] jmpDelta)
                     reader.Seek(jmpBase + jmpDelta[i]);
             }
         }
+
+        private void InterpretNewArray(int token)
+        {
+            int length = 0;
+            StackItem stackItem = PopWithValidation();
+
+            switch (stackItem.Kind)
+            {
+                case StackValueKind.Int32:
+                    length = stackItem.AsInt32();
+                    break;
+                case StackValueKind.NativeInt:
+                    length = (int)stackItem.AsNativeInt();
+                    break;
+                default:
+                    ThrowHelper.ThrowInvalidProgramException();
+                    break;
+            }
+
+            Debug.Assert(length >= 0);
+
+            TypeDesc arrayType = ((TypeDesc)_methodIL.GetObject(token)).MakeArrayType();
+
+            // TODO: Add support for arbitrary  non-primitive types
+            Array array = RuntimeAugments.NewArray(arrayType.GetRuntimeTypeHandle(), length);
+
+            _stack.Push(StackItem.FromObjectRef(array));
+        }
+
+        private void InterpretLoadLength()
+        {
+            Array array = (Array)PopWithValidation().AsObjectRef();
+            _stack.Push(StackItem.FromInt32(array.Length));
+        }
+
+        private void InterpretStoreElement(ILOpcode opcode)
+        {
+            StackItem valueItem = PopWithValidation();
+            StackItem indexItem = PopWithValidation();
+            Array array = (Array)PopWithValidation().AsObjectRef();
+
+            int index = 0;
+            switch (indexItem.Kind)
+            {
+                case StackValueKind.Int32:
+                    index = indexItem.AsInt32();
+                    break;
+                case StackValueKind.NativeInt:
+                    index = (int)indexItem.AsNativeInt();
```

Comment:
This needs bounds check.

Reply:
Do you mean the conversion from `IntPtr` to `int` or the the value of `index` relating to the array length (I already have this in my local branch)?

Reply:
Both of them.